File: app/javascript/channels/api/v1/race_channel.js
const moment = require('moment')

import consumer from '../../consumer'
import { createAlert, showButton, hideButton } from '../../../dom_helpers'
import { ts } from '../../../time'

let raceSubscription
let isConnected = false

document.addEventListener('turbolinks:load', () => {
  if (!window.gon || !window.gon.race) {
    return
  }

  raceSubscription = consumer.subscriptions.create({
    channel:    'Api::V1::RaceChannel',
    race_id:    window.gon.race.id,
    race_type:  window.gon.race.type,
    join_token: window.gon.race.join_token
  }, {
    connected() {
      isConnected = true
      const alert = document.getElementById('disconnected-alert')
      if (alert) {
        alert.parentNode.removeChild(alert)
      }

      const forfeitTime = localStorage.getItem(`race-${window.gon.race.id}-forfeit-time`)
      if (forfeitTime) {
        this.forfeit(parseFloat(forfeitTime))
        localStorage.removeItem(`race-${window.gon.race.id}-forfeit-time`)
      }

      const doneTime = localStorage.getItem(`race-${window.gon.race.id}-done-time`)
      if (doneTime) {
        this.done(parseFloat(doneTime))
        localStorage.removeItem(`race-${window.gon.race.id}-done-time`)
      }
    },

    disconnected() {
      isConnected = false
      const alert = createAlert('danger', 'Disconnected')
      alert.setAttribute('id', 'disconnected-alert')
      document.getElementById('alerts').appendChild(alert)
    },

    received(data) {
      switch(data.type) {
        case 'race_entrants_updated':
          document.getElementById('entrants-table').innerHTML = data.data.entrants_html
          document.getElementById('stats-box').innerHTML = data.data.stats_html
          break;

        case 'race_join_success':
          hideButton('btn-race-join')
          showButton('btn-race-leave')
          showButton('btn-race-ready')
          break;
        case 'race_leave_success':
          hideButton('btn-race-leave')
          showButton('btn-race-join')
          hideButton('btn-race-ready')
          hideButton('btn-race-unready')
          break;
        case 'race_ready_success':
          hideButton('btn-race-ready')
          showButton('btn-race-unready')
          break;
        case 'race_unready_success':
          showButton('btn-race-ready')
          hideButton('btn-race-unready')
          break;

        case 'race_read_only':
          const readAlert = createAlert('warning', 'You will not be able to join this race')
          document.getElementById('alerts').appendChild(readAlert)
          setTimeout(() => { readAlert.parentNode.removeChild(readAlert) }, 5000)
          break;
        case 'race_start_scheduled':
          document.getElementById('stats-box').innerHTML = data.data.stats_html
          hideButton('btn-race-join')
          hideButton('btn-race-leave')
          hideButton('btn-race-ready')
          hideButton('btn-race-unready')
          showButton('btn-race-done')
          showButton('btn-race-forfeit')
          const duration = moment.duration(moment(data.data.race.started_at).valueOf() - ts.now())
          const startAlert = createAlert('warning', `Race starting in ${duration.format('s [seconds]')}!`)
          startAlert.setAttribute('data-start', data.data.race.started_at)
          document.getElementById('alerts').appendChild(startAlert)
          countdown(startAlert)
          break;
        case 'race_done_success':
          hideButton('btn-race-done')
          hideButton('btn-race-forfeit')
          showButton('btn-race-rejoin')
          break;
        case 'race_forfeit_success':
          hideButton('btn-race-done')
          hideButton('btn-race-forfeit')
          showButton('btn-race-rejoin')
          break;
        case 'race_rejoin_success':
          showButton('btn-race-done')
          showButton('btn-race-forfeit')
          hideButton('btn-race-rejoin')
          break;

        case 'race_ended':
          document.getElementById('entrants-table').innerHTML = data.data.entrants_html
          document.getElementById('stats-box').innerHTML = data.data.stats_html
          hideButton('btn-race-forfeit')
          hideButton('btn-race-done')
          hideButton('btn-race-rejoin')
          break;

        case 'new_message':
          document.getElementById('input-list-item').insertAdjacentHTML('afterend', data.data.chat_html)
          break;

        case 'new_attachment':
          document.getElementById('attachments').innerHTML = data.data.attachments_html
          break;

        case 'race_started_error':
        case 'race_join_error':
        case 'in_race_error':
        case 'race_leave_error':
        case 'race_ready_error':
        case 'race_unready_error':
        case 'race_not_started_error':
        case 'race_finished_error':
        case 'race_done_error':
        case 'race_forfeit_error':
        case 'race_rejoin_error':
        case 'message_creation_error':
        case 'fatal_error':
          const errorAlert = createAlert('danger', `Error: ${data.data.message}`)
          document.getElementById('alerts').appendChild(errorAlert)
          setTimeout(() => { errorAlert.parentNode.removeChild(errorAlert) }, 5000)
          break;
      }
    },

    join() {
      this.perform('join')
    },

    leave() {
      this.perform('leave')
    },

    ready() {
      this.perform('ready')
    },

    unready() {
      this.perform('unready')
    },

    forfeit(time) {
      if (isConnected) {
        if (!time) {
          time = ts.now()
        }
        this.perform('forfeit', { server_time: time })
      } else {
        localStorage.setItem(`race-${window.gon.race.id}-forfeit-time`, ts.now())
      }
    },

    done(time) {
      if (isConnected) {
        if (!time) {
          time = ts.now()
        }
        this.perform('done', { server_time: time })
      } else {
        localStorage.setItem(`race-${window.gon.race.id}-done-time`, ts.now())
      }
    },

    rejoin() {
      this.perform('rejoin')
    },

    sendMessage(message) {
      this.perform('send_message', { body: message })
    }
  })

  document.addEventListener('turbolinks:visit', () => {
    consumer.subscriptions.remove(raceSubscription)
  }, {once: true})
})

document.addEventListener('click', (event) => {
  if (event.target.matches('#btn-race-join')) {
    raceSubscription.join()
  }
  if (event.target.matches('#btn-race-leave')) {
    raceSubscription.leave()
  }
  if (event.target.matches('#btn-race-ready')) {
    raceSubscription.ready()
  }
  if (event.target.matches('#btn-race-unready')) {
    raceSubscription.unready()
  }
  if (event.target.matches('#btn-race-done')) {
    raceSubscription.done()
  }
  if (event.target.matches('#btn-race-forfeit')) {
    raceSubscription.forfeit()
  }
  if (event.target.matches('#btn-race-rejoin')) {
    raceSubscription.rejoin()
  }

  if (event.target.matches('#btn-chat-submit')) {
    submitChatInput()
  }
})

document.addEventListener('keypress', (event) => {
  if (event.keyCode !== 13 || event.shiftKey) {
    return
  }

  const chatInput = document.getElementById('input-chat-text')
  if (document.activeElement !== chatInput) {
    return
  }

  submitChatInput()
})

const submitChatInput = () => {
  const chatInput = document.getElementById('input-chat-text')
  if (chatInput.value === '') {
    return
  }

  raceSubscription.sendMessage(chatInput.value)
  chatInput.value = ''
}

const countdown = (alert) => {
  const interval = setInterval(() => {
    const duration = moment.duration(moment(alert.dataset.start).valueOf() - ts.now())
    if (duration.asSeconds() <= 0) {
      clearInterval(interval)
      alert.innerText = 'Race Started!'
      alert.classList.remove('alert-warning')
      alert.classList.add('alert-success')
      setTimeout(() => { alert.parentNode.removeChild(alert) }, 10000)
      return
    }

    alert.innerText = `Race starting in ${duration.format('s [seconds]')}!`
  }, 100)
}

File: app/channels/api/v1/race_channel.rb
class Api::V1::RaceChannel < ApplicationCable::Channel
  def subscribed
    @race = Raceable.race_from_type(params[:race_type]).find_by(id: params[:race_id])
    @permitted = true
    if @race.nil?
      transmit_user('race_not_found', "No race found with id: #{params[:race_id]}")
      reject
      return
    end

    if @race.invite_only_visibility? && !@race.joinable?(user: current_user, token: params[:join_token])
      @permitted = false
      transmit_user('race_read_only', 'Invalid or no join token')
    end

    if @race.secret_visibility? && !@race.joinable?(user: current_user, token: params[:join_token])
      transmit_user('race_invalid_join_token', 'The join token provided is not valid for this race')
      reject
    else
      stream_for(@race)
      return unless params[:state] == '1'

      ws_msg = Api::V1::WebsocketMessage.new(
        'race_state',
        message: 'Race state',
        race:    Api::V4::RaceBlueprint.render_as_hash(@race, view: @race.type, chat: true)
      )

      transmit(Api::V1::WebsocketMessageBlueprint.render_as_hash(ws_msg))
    end
  end

  def unsubscribed
    stop_all_streams
  end

  def join
    return unless @permitted

    update_race_instance

    entrant = @race.entrants.create(user: current_user)
    if entrant.persisted?
      transmit_user('race_join_success', 'Race successfully joined')
      broadcast_race_update('race_entrants_updated', 'A new entrant has join the race')
      GlobalRaceUpdateJob.perform_later('race_entrants_updated', 'A new entrnat has joined', @race)
    else
      transmit_user(entrant.error_status! || 'race_join_error', entrant.errors.full_messages.to_sentence)
    end
  rescue StandardError => e
    Rails.logger.error(e)
    Rollbar.error(e, 'Uncaught error for Api::V1::RaceChannel#join')
    transmit_user('fatal_error', 'A fatal error occurred while processing your message')
  end

  def leave
    return unless @permitted

    update_race_instance

    entrant = Entrant.find_by(raceable: @race, user: current_user)
    return if entrant.nil?

    if entrant.destroy
      transmit_user('race_leave_success', 'Race successfully left')
      broadcast_race_update('race_entrants_updated', 'An entrant has left the race')
      GlobalRaceUpdateJob.perform_later('race_entrants_updated', 'An entrant has left', @race)
    else
      transmit_user(entrant.error_status! || 'race_leave_error', entrant.errors.full_messages.to_sentence)
    end
  rescue StandardError => e
    Rails.logger.error(e)
    Rollbar.error(e, 'Uncaught error for Api::V1::RaceChannel#leave')
    transmit_user('fatal_error', 'A fatal error occurred while processing your message')
  end

  def ready
    return unless @permitted

    update_race_instance

    entrant = Entrant.find_by(raceable: @race, user: current_user)
    return if entrant.nil?

    if entrant.update(readied_at: Time.now.utc)
      transmit_user('race_ready_success', 'Entrant ready successful')
      broadcast_race_update('race_entrants_updated', 'An entrant has readied up')
      maybe_start_race
    else
      transmit_user(entrant.error_status! || 'race_ready_error', entrant.errors.full_messages.to_sentence)
    end
  rescue StandardError => e
    Rails.logger.error(e)
    Rollbar.error(e, 'Uncaught error for Api::V1::RaceChannel#ready')
    transmit_user('fatal_error', 'A fatal error occurred while processing your message')
  end

  def unready
    return unless @permitted

    update_race_instance

    entrant = Entrant.find_by(raceable: @race, user: current_user)
    return if entrant.nil?

    if entrant.update(readied_at: nil)
      transmit_user('race_unready_success', 'Entrant unready successful')
      broadcast_race_update('race_entrants_updated', 'An entrant has unreadied')
    else
      transmit_user(entrant.error_status! || 'race_unready_error', entrant.errors.full_messages.to_sentence)
    end
  rescue StandardError => e
    Rails.logger.error(e)
    Rollbar.error(e, 'Uncaught error for Api::V1::RaceChannel#unready')
    transmit_user('fatal_error', 'A fatal error occurred while processing your message')
  end

  def forfeit(data)
    # Immediately take a timestamp in case there is no server time passed in
    # This is to try and make sure ff's have the most accurate time
    forfeit_time = Time.now.utc
    forfeit_time = Time.at(data['server_time'] / 1000).utc if data['server_time'].present?

    return unless @permitted

    update_race_instance

    entrant = Entrant.find_by(raceable: @race, user: current_user)
    return if entrant.nil?

    if entrant.update(finished_at: nil, forfeited_at: forfeit_time)
      transmit_user('race_forfeit_success', 'Entrant forfeit successful')
      broadcast_race_update('race_entrants_updated', 'An entrant has forfeited')
      maybe_end_race
    else
      transmit_user(entrant.error_status! || 'race_forfeit_error', entrant.errors.full_messages.to_sentence)
    end
  rescue StandardError => e
    Rails.logger.error(e)
    Rollbar.error(e, 'Uncaught error for Api::V1::RaceChannel#forfeit')
    transmit_user('fatal_error', 'A fatal error occurred while processing your message')
  end

  def done(data)
    # Immediately take a timestamp in case there is no server time passed in
    # This is to try and make sure dones have the most accurate time
    done_time = Time.now.utc
    done_time = Time.at(data['server_time'] / 1000).utc if data['server_time'].present?

    return unless @permitted

    update_race_instance

    entrant = Entrant.find_by(raceable: @race, user: current_user)
    return if entrant.nil?

    if entrant.update(finished_at: done_time, forfeited_at: nil)
      transmit_user('race_done_success', 'Entrant done successful')
      broadcast_race_update('race_entrants_updated', 'An entrant has finished')
      maybe_end_race
    else
      transmit_user(entrant.error_status! || 'race_done_error', entrant.errors.full_messages.to_sentence)
    end
  rescue StandardError => e
    Rails.logger.error(e)
    Rollbar.error(e, 'Uncaught error for Api::V1::RaceChannel#done')
    transmit_user('fatal_error', 'A fatal error occurred while processing your message')
  end

  def rejoin
    return unless @permitted

    update_race_instance

    entrant = Entrant.find_by(raceable: @race, user: current_user)
    return if entrant.nil?

    if entrant.update(finished_at: nil, forfeited_at: nil)
      transmit_user('race_rejoin_success', 'Entrant rejoin successful')
      broadcast_race_update('race_entrants_updated', 'An entrant has rejoined the race')
    else
      transmit_user(entrant.error_status! || 'race_rejoin_error', entrant.errors.full_messages.to_sentence)
    end
  rescue StandardError => e
    Rails.logger.error(e)
    Rollbar.error(e, 'Uncaught error for Api::V1::RaceChannel#rejoin')
    transmit_user('fatal_error', 'A fatal error occurred while processing your message')
  end

  def send_message(data)
    return unless @permitted

    update_race_instance
    chat_message = @race.chat_messages.create(
      user: current_user, body: data['body'],
      entrant: @race.entrant_for_user(current_user).present?
    )
    if chat_message.persisted?
      transmit_user('message_creation_success', 'Messages successfully created')
      message = {
        message:      'A new message has been posted',
        chat_message: Api::V4::ChatMessageBlueprint.render_as_hash(chat_message)
      }
      if onsite
        message[:chat_html] = ApplicationController.render(
          partial: 'chat_messages/show',
          locals:  {chat_message: chat_message}
        )
      end

      broadcast_to(
        @race,
        Api::V1::WebsocketMessageBlueprint.render_as_hash(
          Api::V1::WebsocketMessage.new('new_message', message)
        )
      )
    else
      transmit_user('message_creation_error', message.errors.full_messages.to_sentence)
    end
  rescue StandardError => e
    Rails.logger.error(e)
    Rollbar.error(e, 'Uncaught error for Api::V1::RaceChannel#send_message')
    transmit_user('fatal_error', 'A fatal error occurred while processing your message')
  end

  private

  def update_race_instance
    # Instance variables do not update automatically, so we call this function before anything that needs
    # to check the state of the race variable to make sure it isn't stale
    @race.reload
  end

  def transmit_user(type, msg)
    ws_msg = Api::V1::WebsocketMessage.new(type, message: msg)
    transmit(Api::V1::WebsocketMessageBlueprint.render_as_hash(ws_msg))
  end

  def broadcast_race_update(type, msg)
    update_race_instance
    msg = {
      message: msg,
      race:    Api::V4::RaceBlueprint.render_as_hash(@race, view: @race.type)
    }
    if onsite
      msg[:entrants_html] = ApplicationController.render(partial: 'races/entrants_table', locals: {race: @race})
      msg[:stats_html] = ApplicationController.render(partial: 'races/stats', locals: {race: @race})
    end

    ws_msg = Api::V1::WebsocketMessage.new(type, msg)
    broadcast_to(@race, Api::V1::WebsocketMessageBlueprint.render_as_hash(ws_msg))
  end

  # Starts the race if every entrant is readied up, otherwise does nothing
  def maybe_start_race
    update_race_instance
    return if @race.started? || !@race.entrants.all?(&:ready?) || @race.entrants.count < 2

    @race.update(started_at: Time.now.utc + 20.seconds, status: :in_progress)
    broadcast_race_update('race_start_scheduled', 'Race starting soon')
    GlobalRaceUpdateJob.perform_later('race_started', 'A race has started', @race)
  end

  # Ends the race if all entrants have either finished or forfeited, otherwise does nothing
  def maybe_end_race
    update_race_instance
    return if !@race.started? || !@race.entrants.all?(&:done?)

    @race.update(status: :ended)
    broadcast_race_update('race_ended', 'All entrants have finished')
    GlobalRaceUpdateJob.perform_later('race_finished', 'A race has finished', @race)
  end
end
